iptables: handle case where kernel IPv6 support is disabled

jibi opened the pull request:
Currently, even if the kernel IPv6 support is disabled (i.e. the
ipv6.disable=1 kernel parameter is set), the agent will try to find or
load the ip6tables kernel modules, and if successful it will assume
ip6tables is supported.

This will result in a fatal error since all ip6tables commands will
fail with "Address family not supported".

This commit fixes this by setting the haveIp6tables IptablesManager's
property to false in case the kernel does not support IPv6.

Fixes: d812b925de ("iptables: don't ignore errors")
Fixes: #18513
Fixes: #20672

julianwiedmann commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -297,6 +297,18 @@ func (m *IptablesManager) Init() {
 		haveIp6tables = false
 	}
+
+	ipv6Disabled, err := sysctl.Read("net.ipv6.conf.all.disable_ipv6")
```
I'm a bit uneasy about [this part](https://github.com/torvalds/linux/blob/6e7765cb477a9753670d4351d14de93f1e9dbbd4/Documentation/networking/ip-sysctl.rst?plain=1#L1998):

> **Reading this value does not have any particular meaning**. It does not say
> whether IPv6 support is enabled or disabled. Returned value can be 1
> also in the case when some interface has ``disable_ipv6`` set to 0 and
> has configured IPv6 addresses.

If this is strictly about the `ipv6.disable=1` option, should we be looking at `/sys/module/ipv6/parameters/disable` instead?

jibi replied:
So in my test I set the `ipv6.disable=1` kernel parameter (which is what the bug report is about) which then results in `net.ipv6.conf.all.disable_ipv6` being set to `1`.

I'm going for this check mostly for consistency as we are already using it in another part of the codebase:

https://github.com/cilium/cilium/blob/94ea7fc93e0af5bf3ddab5b67508003ca47678cd/pkg/datapath/loader/base.go#L316

but relying on `/sys/module/ipv6/parameters/disable` would be fine as well

julianwiedmann replied:
> So in my test I set the `ipv6.disable=1` kernel parameter (which is what the bug report is about) which then results in `net.ipv6.conf.all.disable_ipv6` being set to `1`.
> 
> I'm going for this check mostly for consistency as we are already using it in another part of the codebase:
> 
> https://github.com/cilium/cilium/blob/94ea7fc93e0af5bf3ddab5b67508003ca47678cd/pkg/datapath/loader/base.go#L316
> 
> but relying on `/sys/module/ipv6/parameters/disable` would be fine as well


The sysctl doesn't feel like a good fit for catching `ipv6.disable=1` to me. It just controls IPv6 operations on the selected network interface(s), no? For instance, if I manually set `net.ipv6.conf.all.disable_ipv6=1` in combination with `ipv6.disable=0`, then running the reproducer (`ip6tables -w 5 -t nat -S`) still works fine.

jibi replied:
Right, setting `net.ipv6.conf.all.disable_ipv6=1` will not impact the `ip6tables` command, although the inet6 addresses get removed from the interfaces:
```
vagrant@k8s1:~$ ip a | grep inet6
vagrant@k8s1:~$
```

let's switch to `/sys/module/ipv6/parameters/disable` then :+1:

julianwiedmann reviewed (changes requested):
ugh, I remember having an uneasy feeling about this...

For the commit message:
- can you please add what specific case this covers? It's the [ipv6.disable=1](https://github.com/torvalds/linux/blob/master/Documentation/networking/ipv6.rst) module parameter, correct?
- think we need a `Fixes:` tag for the commit that introduced the regression.

julianwiedmann commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -297,6 +298,18 @@ func (m *IptablesManager) Init() {
+		log.Debug(
+			"Kernel does not support IPv6, disabling IPv6 iptables support")
+		haveIp6tables = false
+	}
```
Wondering whether we should `Fatal()` if `option.Config.EnableIPv6` is set, to match what we do in the earlier check for the ip6_tables kernel modules.

jibi replied:
Yeah good shout :+1:

julianwiedmann reviewed (approved):
lgtm, thanks for the quick fix jibi! One further comment, but feel free to ignore.

jibi committed:
iptables: handle case where kernel IPv6 support is disabled

Currently, even if the kernel IPv6 support is disabled (i.e. the
ipv6.disable=1 kernel parameter is set), the agent will try to find or
load the ip6tables kernel modules, and if successful it will assume
ip6tables is supported.

This will result in a fatal error since all ip6tables commands will
fail with "Address family not supported".

This commit fixes this by setting the haveIp6tables IptablesManager's
property to false in case the kernel does not support IPv6.

Fixes: d812b925de ("iptables: don't ignore errors")
Fixes: #18513
Fixes: #20672
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

YutaroHayakawa commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -297,6 +298,22 @@ func (m *IptablesManager) Init() {
+			"Unable to read /sys/module/ipv6/parameters/disable, disabling IPv6 iptables support")
+		haveIp6tables = false
+	} else if strings.TrimSuffix(string(ipv6Disabled), "\n") == "1" {
+		log.Debug(
```
I think we should show this message to users even if it is not a Fatal. Maybe it should be a Warning?

jibi replied:
I'm following the same pattern as above (i.e. emitting a debug logline in case we fail to find/load ip6tables modules):
https://github.com/cilium/cilium/blob/78d075d1bedf7f6e3890dd2a9fab20383201f3c4/pkg/datapath/iptables/iptables.go#L290-L297

YutaroHayakawa reviewed (approved):
LGTM with a nit (almost the same as Julian's one) 👍
